### ISSUE TITLE
Gradients wrt LinearOperators on scalar functions not matching expectations

### ISSUE THREAD

**Issue by quattro** [patrick-kidger/lineax#144]
Hi 👋 ,

Thanks again for the wonderful library. Currently `jax.grad` (or `equinox.filter_grad`) wrt `lineax.AbstractLinearOperator` arguments dont return the expected gradients. I'm not sure if this is out-of-scope, given the focus on solvers/solving systems of linear equations, but at least thought it'd be worth documenting.

In the below example, the expected gradient output should be an `lineax.AbstractLinearOperator` whose `*.as_matrix()` returns they identity. However, this isn't the case. When the argument is a `lineax.MatrixLinearOperator` it seems to work correctly, but anything more complex than that (e.g., `lineax.ComposedLinearOperator`, etc) fails.

```python
import jax
import jax.numpy as jnp
import jax.random as rdm
import lineax as lx
import equinox as eqx

seed = 0
key = rdm.PRNGKey(seed)
N, P = 5, 3
X = rdm.normal(key, shape=(N, P))
M = lx.MatrixLinearOperator(X)
A = M.T @ M

def foo(A): 
  return jnp.trace(A.as_matrix())

jax.grad(foo)(A)
jax.grad(foo)(A).as_matrix()
jax.grad(jnp.trace)(A.as_matrix())                    # works!

# convert to matrix then reconvert back to linop
AA = lx.MatrixLinearOperator(A.as_matrix())
jax.grad(jnp.trace)(AA.as_matrix())                   # works!
jax.grad(foo)(AA).as_matrix()                         # works!


jax.grad(foo)(2. * AA).as_matrix()
jax.grad(foo)(AA + AA ).as_matrix()

eqx.filter_grad(foo)(2. * AA).as_matrix()
eqx.filter_grad(foo)(AA + AA).as_matrix()
```

**Comment by patrick-kidger** [comment]
Thanks for the issue! It's definitely nonintuitive the first time one comes across this, but this is actually expected.

What's going on is that `jax.grad` takes derivatives with respect to the arrays in the pytree structure -- the way that those arrays are used in the primal determines what value that the corresponding gradient arrays take... and then those just so happened to be organized into a pytree that is a linear operator.

To help elucidate this, here is an example. In this case we only ever have a single differentiable parameter, which is the `1.0` we multiply by. Each of our gradients likewise consists of just a single scalar value (wrapped up into a pytree):
```python
import jax
import jax.numpy as jnp
import lineax as lx


def fn(op):
  return jnp.trace(op.as_matrix())

for n in range(1, 6):
    struct = jax.ShapeDtypeStruct((n,), jnp.float32)
    op = 1.0 * lx.IdentityLinearOperator(struct)
    grad_op = jax.grad(fn)(op)
    print(grad_op.as_matrix())

# [[1.]]
# [[2. 0.]
#  [0. 2.]]
# [[3. 0. 0.]
#  [0. 3. 0.]
#  [0. 0. 3.]]
# [[4. 0. 0. 0.]
#  [0. 4. 0. 0.]
#  [0. 0. 4. 0.]
#  [0. 0. 0. 4.]]
# [[5. 0. 0. 0. 0.]
#  [0. 5. 0. 0. 0.]
#  [0. 0. 5. 0. 0.]
#  [0. 0. 0. 5. 0.]
#  [0. 0. 0. 0. 5.]]
```
It's a mistake to ever call `.as_matrix()` (or to perform any other linear operator like function) on the gradient. Gradients are pytrees of arrays, nothing else.

I definitely agree this is pretty surprising the first time you see it. I'd be happy to take a suggestion / a PR on where you'd have found it useful to update our docs on this!
